src/libraries/Common/src/Roslyn/SyntaxValueProvider_ForAttributeWithMetadataName.cs

```diff
@@ -0,0 +1,188 @@
+        // a single file with a ton of matching nodes in it).
+        var groupedNodes = collectedNodes.SelectMany(
+            static (array, cancellationToken) =>
+                array.GroupBy(static n => n.SyntaxTree)
+                     .Select(static g => new SyntaxNodeGrouping<SyntaxNode>(g)))/*.WithTrackingName("groupedNodes_ForAttributeWithMetadataName")*/;
+
+        var compilationAndGroupedNodesProvider = groupedNodes
+            .Combine(context.CompilationProvider)
+            /*.WithTrackingName("compilationAndGroupedNodes_ForAttributeWithMetadataName")*/;
+
+        var syntaxHelper = CSharpSyntaxHelper.Instance;
+        var finalProvider = compilationAndGroupedNodesProvider.SelectMany((tuple, cancellationToken) =>
+        {
+            var (grouping, compilation) = tuple;
+
+            var result = new ValueListBuilder<T>(Span<T>.Empty);
+            var syntaxTree = grouping.SyntaxTree;
+            var semanticModel = compilation.GetSemanticModel(syntaxTree);
+
+            foreach (var targetNode in grouping.SyntaxNodes)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var targetSymbol =
+                    targetNode is ICompilationUnitSyntax compilationUnit ? semanticModel.Compilation.Assembly :
+                    syntaxHelper.IsLambdaExpression(targetNode) ? semanticModel.GetSymbolInfo(targetNode, cancellationToken).Symbol :
+                    semanticModel.GetDeclaredSymbol(targetNode, cancellationToken);
+                if (targetSymbol is null)
+                    continue;
+
+                var attributes = getMatchingAttributes(targetNode, targetSymbol, fullyQualifiedMetadataName);
+                if (attributes.Length > 0)
+                {
+                    result.Append(transform(
+                        new GeneratorAttributeSyntaxContext(targetNode, targetSymbol, semanticModel, attributes),
+                        cancellationToken));
+                }
+            }
+
+            return result.AsSpan().ToImmutableArray();
+        })/*.WithTrackingName("result_ForAttributeWithMetadataName")*/;
+
+        return finalProvider;
+
+        static ImmutableArray<AttributeData> getMatchingAttributes(
+            SyntaxNode attributeTarget,
+            ISymbol symbol,
+            string fullyQualifiedMetadataName)
+        {
+            var targetSyntaxTree = attributeTarget.SyntaxTree;
+            var result = new ValueListBuilder<AttributeData>(Span<AttributeData>.Empty);
+
+            addMatchingAttributes(ref result, symbol.GetAttributes());
+            addMatchingAttributes(ref result, (symbol as IMethodSymbol)?.GetReturnTypeAttributes());
+
+            if (symbol is IAssemblySymbol assemblySymbol)
+            {
+                foreach (var module in assemblySymbol.Modules)
+                    addMatchingAttributes(ref result, module.GetAttributes());
+            }
+
+            return result.AsSpan().ToImmutableArray();
```

Comment:
Where is `result` being disposed in order to return the pooled resources?

Reply:
oh!  i didn't know it has disposal semantics.  My bad!

Reply:
fixed.